server/src/main/java/game/ai/bot/BotFactory.java

```diff
@@ -1,22 +1,93 @@
 package game.ai.bot;
 
+import game.Game;
 import game.world.World;
 
+import java.util.LinkedList;
+import java.util.List;
 import java.util.Random;
 
 public class BotFactory {
-    private static final Random random = new Random();
 
-    public static Bot createBot(World world) {
-        final var p = random.nextDouble();
+    /**
+     * Creates n of the various Bot types with the type chosen
+     * depending on the current bot population and the target population such that bot distribution
+     * converges towards the target bot distribution
+     */
+    public static List<Bot> createBotsDynamically(World world, Game game, int numberOfBots) {
 
-        if (p < 0.1) {
-            return new ScaredBot(world);
-        } else if (p < 0.4) {
-            return new KamikazeBot(world);
-        } else if (p < 0.8) {
-            return new HungryBot(world);
+        final List<Bot> botList = new LinkedList<>();
+
+        int scaredBotCount = game.countBotsOfType(ScaredBot.class);
+        int kamikazeBotCount = game.countBotsOfType(KamikazeBot.class);
+        int stupidBotCount = game.countBotsOfType(StupidBot.class);
+
+        double scaredBotTargetCount = game.config.botPopulationDistributionTarget.scaredBotTarget;
+        double kamikazeBotTargetCount = game.config.botPopulationDistributionTarget.kamikazeBotTarget;
+        double stupidBotTargetCount = game.config.botPopulationDistributionTarget.stupidBotTarget;
+
+        int botCount = game.getNumberOfBots();
+        assert (scaredBotCount + kamikazeBotCount + stupidBotCount == botCount);
+
+        //I know, I know...but it should work and avoid division by Zero if a target is 0.
+        if (scaredBotTargetCount == 0) {scaredBotTargetCount = 0.1;}
+        if (kamikazeBotTargetCount == 0) {kamikazeBotTargetCount = 0.1;}
+        if(stupidBotTargetCount == 0){stupidBotTargetCount = 0.1;}
+
```

Comment:
Yeah no, unacceptable. Instead I suggest you write a method for computing the quota based on count and target count (both ints) and a special case for target quota being 0 (either return 0.0 or 1.0)